File: python/math/floor_sum.py
# computes sum((a*i+b)//m for i in range(n)) in O(log m)
# 0 <= n; 1 <= m
def floor_sum(n, m, a, b):
    assert 0 <= n and 1 <= m
    ans = 0
    if a >= m or a < 0: ans+= (n-1)*n//2 * (a//m); a%= m
    if b >= m or b < 0: ans+= n * (b//m); b%= m
    if a == 0: return ans
    y, z = divmod(a*n+b, m)
    return ans + floor_sum(y, a, m, z)

File: cpp/math/floor_sum.cpp
template <typename T>
T floor_sum(T n, T m, T a, T b){
	assert(0 <= n && 1 <= m);
	T ans = 0;
	if(a >= m || a < 0) ans+= (n-1)*n/2 * (a/m), a%= m;
	if(b >= m || b < 0) ans+= n * (b/m), b%= m;
	if(a == 0) return ans;
	ll y = (a*n+b)/m, z = (a*n+b)%m;
	return ans + floor_sum(y, a, m, z);
}
